tests/Constraints/CoerciveTest.php

```diff
@@ -11,271 +11,201 @@
+        $tests[] = array(
+            '{"properties":{"propertyOne":{"type":["number", "string"]}}}',
+            '{"propertyOne":42}',
+            'integer', 'integer', 42, true
+        );
 
-        $this->assertTrue($value->number === 1.5);
-        $this->assertTrue($value->integer === 1);
-        $this->assertTrue($value->negativeInteger === -2);
-        $this->assertTrue($value->boolean === true);
+        // #41 check multiple types (last valid)
+        $tests[] = array(
+            '{"properties":{"propertyOne":{"type":["number", "string"]}}}',
+            '{"propertyOne":"42"}',
+            'string', 'string', '42', true
+        );
 
-        $this->assertTrue($validator->isValid(), print_r($validator->getErrors(), true));
-    }
+        // #42 check the meaning of life
+        $tests[] = array(
+            '{"properties":{"propertyOne":{"type":"any"}}}',
+            '{"propertyOne":"42"}',
+            'string', 'string', '42', true
+        );
 
-    /**
-     * @dataProvider getInvalidCoerceTests
-     */
-    public function testInvalidCoerceCases($input, $schema, $errors = array())
-    {
-        $checkMode = Constraint::CHECK_MODE_TYPE_CAST | Constraint::CHECK_MODE_COERCE_TYPES;
+        // #43 check turple coercion
+        $tests[] = array(
+            '{"properties":{"propertyOne":{"type":"array","items":[{"type":"number"},{"type":"string"}]}}}',
+            '{"propertyOne":["42", 42]}',
+            'array', 'array', array(42, '42'), true
+        );
 
-        $schemaStorage = new SchemaStorage($this->getUriRetrieverMock(json_decode($schema)));
-        $schema = $schemaStorage->getSchema('http://www.my-domain.com/schema.json');
+        // #44 check early coercion
+        $tests[] = array(
+            '{"properties":{"propertyOne":{"type":["object", "number", "string"]}}}',
+            '{"propertyOne":"42"}',
+            'string', 'integer', 42, true, true
+        );
 
-        $validator = new Validator(new Factory($schemaStorage, null, $checkMode));
-        $value = json_decode($input);
-        $validator->validate($value, $schema, $checkMode);
+        // #45 check multiple types (none valid)
+        $tests[] = array(
+            '{"properties":{"propertyOne":{"type":["number", "boolean"]}}}',
+            '{"propertyOne":"42"}',
+            'string', 'integer', 42, true
+        );
 
-        if (array() !== $errors) {
-            $this->assertEquals($errors, $validator->getErrors(), print_r($validator->getErrors(), true));
-        }
-        $this->assertFalse($validator->isValid(), print_r($validator->getErrors(), true));
+        return $tests;
     }
 
-    /**
-     * @dataProvider getInvalidCoerceTests
-     */
-    public function testInvalidCoerceCasesUsingAssoc($input, $schema, $errors = array())
+    /** @dataProvider dataCoerceCases **/
+    public function testCoerceCases($schema, $data, $startType, $endType, $endValue, $valid, $early = false, $assoc = false)
```

Comment:
What if the 7th argument was just something like `$flags`? Then instead of passing in `true` you could pass in `CHECK_MODE_EARLY_COERCE`, which would be a little easier to read, and you don't have to keep adding more arguments to this method signature as more flags are added.